APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

robson-sanchez opened the pull request:
- Agent version equal or greater than '1.46.0' should use the Agent Message API v3. Other versions still using Agent Message API v2.

The 2 first commits are part of another PR: #37. If you want to review it first, I could rebase this PR after merging to reduce the number of files to be analyzed. Otherwise, I should close that PR.

robson-sanchez committed:
APP-904 Support 'multipart/form-data' in the HTTP clients

robson-sanchez committed:
APP-905 Agent client should support posting message using new version from Agent API

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

robson-sanchez committed:
APP-905 Agent client should support posting message using new version from Agent API

- Fix issues to post message through the new version from Agent API

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Event producers should remove SNAPSHOT substring from the service version

ecarrenho-daitan commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/MessageApiClient.java:
```diff
@@ -16,67 +16,17 @@
 /**
- * Holds all endpoints to deal with message.
- * Created by rsanchez on 23/02/17.
+ * Interface to be implemented by HTTP clients that would like to post message through the Agent
```
... clients that post...

robson-sanchez replied:
DONE

ecarrenho-daitan commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V2MessageApiClient.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.Map;
+
+/**
+ * Holds all endpoints to deal with message.
```
Add further information related to messageV2

robson-sanchez replied:
DONE

ecarrenho-daitan commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V3MessageApiClient.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.Map;
+
+/**
+ * Holds all endpoints to deal with message related to new MessageML version.
```
Add further information to mention message v3

robson-sanchez replied:
DONE

ecarrenho-daitan commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V3MessageApiClient.java:
```diff
@@ -0,0 +1,92 @@
+      if (message instanceof V3Message) {
+        String entityJSON = ((V3Message) message).getData();
+
+        if (StringUtils.isNotEmpty(entityJSON)) {
```
use isNotBlank

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Select the correct Agent Message API Client according to the MessageML version included in the message object

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/AgentApiClient.java:
```diff
@@ -28,6 +30,7 @@
  * Created by rsanchez on 22/02/17.
  */
 @Component
+@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
```
Why is this a prototype?  It doesn't seem to have any state that changes.  And it looks thread safe.

robson-sanchez replied:
@dnsymphony It's required because we need to instantiate two low-level HTTP clients for each API version because they are using different serializers. One of them needs to serialize JSON objects, the other needs to send requests using 'multipart/form-data'.

The serializer is being kept inside this class, that's the motivation to have more than one instance.

This behavior will have changed in the future when the applications should handle all events using messageML v2.

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/BaseMessageApiClient.java:
```diff
@@ -0,0 +1,51 @@
+ */
+public class BaseMessageApiClient {
+
+  protected void validateParams(String sessionToken, String kmToken, String streamId,
```
Javadoc

robson-sanchez replied:
DONE

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/BaseMessageApiClient.java:
```diff
@@ -0,0 +1,51 @@
+ * Base Message API client.
+ * Created by rsanchez on 27/03/17.
+ */
+public class BaseMessageApiClient {
```
Add abstract
Add implements MessageApiClient

robson-sanchez replied:
DONE

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V2MessageApiClient.java:
```diff
@@ -0,0 +1,66 @@
+ * Holds all endpoints to deal with message.
+ * Created by rsanchez on 23/02/17.
+ */
+public class V2MessageApiClient extends BaseMessageApiClient implements MessageApiClient {
```
Remove implements MessageApiClient

robson-sanchez replied:
DONE

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V3MessageApiClient.java:
```diff
@@ -0,0 +1,88 @@
+ * Holds all endpoints to deal with message related to new MessageML version.
+ * Created by rsanchez on 27/03/17.
+ */
+public class V3MessageApiClient extends BaseMessageApiClient implements MessageApiClient {
```
Remove implements MessageApiClient

robson-sanchez replied:
DONE

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V2MessageApiClient.java:
```diff
@@ -0,0 +1,66 @@
+ */
+public class V2MessageApiClient extends BaseMessageApiClient implements MessageApiClient {
+
+  private static final String SESSION_TOKEN_HEADER_PARAM = "sessionToken";
```
Move to interface or base class

robson-sanchez replied:
DONE

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V3MessageApiClient.java:
```diff
@@ -0,0 +1,88 @@
+
+      return apiClient.doPost(path, headerParams, Collections.<String, String>emptyMap(), multiPart, Message.class);
+    } catch (Exception e) {
+      throw new RemoteApiException(500, e);
```
Add some logging for context - at least log the stream ID.  You could also log the safe version of the session ID.  See implementation SessionUtils.getLogSafeSessionId(sessionIdStr)

robson-sanchez replied:
DONE

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bootstrap/IntegrationBootstrapContext.java:
```diff
@@ -129,11 +138,31 @@ public void initIntegrations() {
+        LOGGER.debug("Polling AGENT health check");
+
+        HealthCheckServiceEvent event = new HealthCheckServiceEvent(AGENT_SERVICE_NAME);
+        publisher.publishEvent(event);
```
Interesting. I didn't think you could do this. I thought only ApplicationEvent object could be published.

dnsymphony replied:
This seems like a lot of indirection.  You've got one thread that sends events.  A listener receives that event and calls Agent health check. Based on results of health check, the listener may fire another event that says version of Agent has changed.  Listener for version changed event then stores the new value.  Why isn't the health check call just it's own recurring runnable?

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bootstrap/IntegrationBootstrapContext.java:
```diff
@@ -129,11 +138,31 @@ public void initIntegrations() {
   }
 
   /**
+   * Schedule to dispatch health-check service event to monitor the Agent version.
```
Add: "This allows the Integration Bridge to detect if the Agent is upgraded while the Integration Bridge is running"

robson-sanchez replied:
DONE

Commit:
Merge branch 'dev' into APP-906

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/IntegrationBridgeExceptionHandler.java:
```diff
@@ -110,7 +110,7 @@ public void handleRemoteApiException(RemoteApiException remoteException,
       updateStreams(instance, integrationUser, stream);
     } else if (Status.BAD_REQUEST.equals(status)) {
-      LOGGER.warn("Invalid messageML: " + message, remoteException);
+      LOGGER.warn("Invalid messageML: " + message.toString(), remoteException);
```
Include information about the integration instance and stream.

robson-sanchez replied:
DONE

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/IntegrationBridgeImpl.java:
```diff
@@ -93,40 +91,17 @@
-    Message messageSubmission = buildMessage(message);
-    Message messageResponse = streamService.postMessage(integrationUser, stream, messageSubmission);
+    Message messageResponse = streamService.postMessage(integrationUser, stream, message);
     LOGGER.info("Message posted to stream {} ", stream);
```
Include user

robson-sanchez replied:
DONE

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/StreamServiceImpl.java:
```diff
@@ -37,44 +43,58 @@
 
   @Autowired
-  private AgentApiClient agentApiClient;
+  private AgentApiClient agentV2ApiClient;
```
Does this work?  How do you autowire two different beans with the same type without specifying bean/resource name?

robson-sanchez replied:
It works because AgentApiClient is a prototype component. I'll have two copies from the AgentApiClient. Each copy will be configured to be used with a specific version of the Message API.

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/StreamServiceImpl.java:
```diff
@@ -37,44 +43,58 @@
+
+    MessageApiClient messageApiClient = new V2MessageApiClient(agentV2ApiClient);
+
+    apiResolver.put(MessageMLVersion.V1, messageApiClient);
```
Same for both? 

Ok.  I see.  It gets overridden below when version updated event is handled.  You should add some comments to explain that.  At first glance, this looks like a bug.

robson-sanchez replied:
DONE

dnsymphony commented on integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/services/ServiceHealthIndicator.java:
```diff
@@ -82,26 +93,48 @@
+
   @PostConstruct
   public void init() {
+    this.currentVersion = null;
```
This is redundant.  Fields are always null when class is created.

robson-sanchez replied:
DONE

dnsymphony commented on integration-core/src/test/java/org/symphonyoss/integration/core/bridge/StreamServiceTest.java:
```diff
@@ -136,25 +169,32 @@ private IntegrationInstance mockInstance() {
     when(authenticationProxy.getToken(INTEGRATION_USER)).thenReturn(
         AuthenticationToken.VOID_AUTH_TOKEN);
-    when(messagesApi.postMessage(anyString(), anyString(), anyString(),
+    when(messageApiClient.postMessage(anyString(), anyString(), anyString(),
```
Avoid using the "any" matchers if you can know the actual values being used

dnsymphony commented on integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/services/AgentHealthIndicator.java:
```diff
@@ -32,6 +35,8 @@
 
   private static final String AGENT_URL_PATH = "/v1/HealthCheck";
 
+  private static final Version AGENT_MESSAGEML_VERSION2 = Version.valueOf("1.46.0");
```
This string is showing up all over the place. Can it be a shared constant?

dnsymphony commented on integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/event/ServiceVersionUpdatedEvent.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * Created by rsanchez on 21/03/17.
+ */
+public class ServiceVersionUpdatedEvent {
```
I'm confused about the difference between this event and the MessageMLVersionUpdatedEvent.  They seem to be triggering the same thing but are inconsistently used.

robson-sanchez replied:
ServiceVersionUpdatedEvent is being triggered when the health check identifies the service version was updated. This event is being used to notify the HTTP clients that a new version of Agent API is available to be used.

MessageMLVersionUpdatedEvent is being triggered exclusively by the Agent health check to notify the parsers the MessageML v2 is available to be used.

Add this info as javadoc on the respective classes.

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/StreamServiceImpl.java:
```diff
@@ -123,4 +145,28 @@ public Stream createIM(String integrationUser, Long userId) throws RemoteApiExce
+   * @param event Service version updated event
+   */
+  @EventListener
+  public void handleServiceVersionUpdatedEvent(ServiceVersionUpdatedEvent event) {
```
I would expect this to be what happens on a MessageMLVersionUpdated event.

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Fix PR comments

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Fix PR comments

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Improve documentation

dnsymphony commented on integration-agent-api-client/src/main/java/org/symphonyoss/integration/agent/api/client/V3MessageApiClient.java:
```diff
@@ -77,7 +77,10 @@ public Message postMessage(String sessionToken, String kmToken, String streamId,
     } catch (Exception e) {
-      throw new RemoteApiException(500, e);
+      String errorMessage =
+          String.format("Fail to post message %s to stream %s due to %s", message.toString(),
```
Don't log message contents!

robson-sanchez replied:
DONE

dnsymphony commented on integration-core/src/main/java/org/symphonyoss/integration/core/bridge/IntegrationBridgeExceptionHandler.java:
```diff
@@ -110,7 +110,9 @@ public void handleRemoteApiException(RemoteApiException remoteException,
     } else if (Status.BAD_REQUEST.equals(status)) {
-      LOGGER.warn("Invalid messageML: " + message.toString(), remoteException);
+      String logMessage = String.format("Invalid message posted to stream %s.\nMessage: "
+          + "%s\nInstance: %s", stream, message.toString(), instance.getInstanceId());
```
Do not log message contents

robson-sanchez replied:
DONE

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Fix unit tests
- Shared constants

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Don't log message content

robson-sanchez committed:
APP-906 Stream Service should select the correct version of Agent API to be used based on Agent version

- Improve documentation

rsanchez-daitan commented:
@dnsymphony @ecarrenho-daitan please check again
